plugins/explore/src/components.cpp

```diff
@@ -0,0 +1,659 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2023 The TEnzir Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include "explore/components.hpp"
+
+#include "explore/elements.hpp"
+#include "explore/ui_state.hpp"
+
+#include <tenzir/collect.hpp>
+#include <tenzir/concept/printable/tenzir/data.hpp>
+#include <tenzir/concept/printable/to_string.hpp>
+#include <tenzir/detail/narrow.hpp>
+#include <tenzir/table_slice.hpp>
+#include <tenzir/table_slice_column.hpp>
+
+#include <ftxui/component/component.hpp>
+#include <ftxui/component/screen_interactive.hpp>
+#include <ftxui/dom/elements.hpp>
+
+#include <unordered_set>
+
+using namespace ftxui;
+
+namespace tenzir::plugins::explore {
+
+auto lift(Element e) -> Component {
+  class Impl : public ComponentBase {
+  public:
+    explicit Impl(Element e) : element_{std::move(e)} {
+    }
+
+    auto Render() -> Element final {
+      return element_;
+    }
+
+  private:
+    Element element_;
+  };
+  return Make<Impl>(std::move(e));
+}
+
+auto enframe(const Component& component) -> Component {
+  return Renderer(component, [component] {
+    return component->Render() | frame;
+  });
+}
+
+namespace {
+
+/// A table showing user help and keyboard shortcuts.
+auto Help() -> Component {
+  auto table = Table({
+    {" Key ", " Alias ", " Description "},
+    {"t", " ", "toggle display of type annotations in headers"},
+    {"n", " ", "toggle display of navigator"},
+    {"k", "↑", "move focus one component up"},
+    {"j", "↓", "move focus one component down"},
+    {"h", "←", "move focus one component to the left"},
+    {"l", "→", "move focus one component to the right"},
+    {"K", " ", "move up in navigator"},
+    {"J", " ", "move down in navigator"},
+    {"H", " ", "move left in navigator"},
+    {"L", " ", "move right in navigator"},
+    {"?", "", "show this help"},
+    {"q", "ESC", "quit the UI"},
+  });
+  table.SelectAll().Border(ROUNDED);
+  table.SelectRow(0).Decorate(bold);
+  table.SelectRow(0).SeparatorHorizontal(ROUNDED);
+  table.SelectRow(0).Border(ROUNDED);
+  table.SelectColumn(0).DecorateCells(center);
+  table.SelectColumn(1).DecorateCells(center);
+  return lift(table.Render());
+}
+
+enum class alignment { left, center, right };
+
+auto colorize(auto make, view<data> value, const struct theme& theme) {
+  auto factory = detail::overload{
+    [&](const auto& x) {
+      return make(x, alignment::left, theme.palette.text);
+    },
+    [&](caf::none_t) {
+      return make("∅", alignment::center, theme.palette.muted);
```

Comment:
I think we should align this with the web-based Explorer, i.e., show "null" as a string and left-align it.